sound/soc/sof/ipc4-pcm.c

```diff
@@ -227,8 +229,201 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 	return 0;
 }
 
+static int sof_ipc4_build_time_info(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm)
+{
+	struct sof_ipc4_llp_reading_slot llp_slot;
+	struct sof_ipc4_timestamp_info *info;
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_dai *dai;
+	int i;
+
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	info = spcm->time_info;
+
+	if (!host_copier || !dai_copier) {
+		dev_err(sdev->dev, "failed to find host or dai copier\n");
+		return -EINVAL;
+	}
+
+	info->host_copier = host_copier;
+	info->dai_copier = dai_copier;
+
+	info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_gpdma_reading_slots) +
+				    sdev->fw_info_box.offset;
+
+	/* find llp slot used by current dai */
+	for (i = 0; i < SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS; i++) {
+		sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+
+		if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+			break;
+
+		info->llp_offset += sizeof(struct sof_ipc4_llp_reading_slot);
+	}
+
+	if (i >= SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS) {
+		/* if no llp gpdma slot is used, check aggregated sdw slot */
+		info->llp_offset = offsetof(struct sof_ipc4_fw_registers, llp_sndw_reading_slots) +
+						sdev->fw_info_box.offset;
+		for (i = 0; i < SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS - 1; i++) {
+			sof_mailbox_read(sdev, info->llp_offset, &llp_slot, sizeof(llp_slot));
+
+			if (llp_slot.node_id == dai_copier->data.gtw_cfg.node_id)
+				break;
+
+			info->llp_offset += sizeof(struct sof_ipc4_llp_reading_slot);
+		}
+	}
+
+	/* no llp support */
+	if (i >= SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS - 1) {
+		dev_warn(sdev->dev, "no llp found, fall back to default path");
+		info->llp_offset = 0;
+	}
+
+	return 0;
+}
+
+static int sof_ipc4_pcm_hw_params(struct snd_soc_component *component,
+				  struct snd_pcm_substream *substream,
+				  struct snd_pcm_hw_params *params,
+				  struct snd_sof_platform_stream_params *platform_params)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_timestamp_info *time_info;
+	struct snd_sof_pcm *spcm;
+	u32 abi_version;
+	u32 abi_offset;
+
+	abi_offset = offsetof(struct sof_ipc4_fw_registers, abi_ver);
+	sof_mailbox_read(sdev, sdev->fw_info_box.offset + abi_offset, &abi_version,
+			 sizeof(abi_version));
+
+	if (abi_version != SOF_IPC4_FW_REGS_ABI_VER) {
+		dev_warn(sdev->dev, "unsupported abi version %d for time stamp\n",
+			 abi_version);
+		return -EINVAL;
+	}
```

Comment:
Why aren't we doing this at probe time? Is it going to change between consecutive playbacks or captures?

Reply:
we need to use old version of fw.  We need to check it when we want to use fw_reg.  Thanks!

Reply:
sorry , missed.  please check the latest code